charmcraft/commands/build.py

```diff
@@ -138,24 +140,79 @@ def __init__(self, args, config):
+    def build_charm(self, bases_config: Optional[BasesConfiguration]) -> str:
+        """Build the charm.
+
+        :param bases_config: Bases configuration to use for build.
+
+        :returns: File name of charm.
+        """
         logger.debug("Building charm in '%s'", self.buildpath)
 
         if self.buildpath.exists():
             shutil.rmtree(str(self.buildpath))
         self.buildpath.mkdir()
 
-        create_manifest(self.buildpath, self.config.project.started_at)
+        create_manifest(self.buildpath, self.config.project.started_at, bases_config)
 
         linked_entrypoint = self.handle_generic_paths()
         self.handle_dispatcher(linked_entrypoint)
         self.handle_dependencies()
-        zipname = self.handle_package()
+        zipname = self.handle_package(bases_config)
 
         logger.info("Created '%s'.", zipname)
         return zipname
 
+    def run(self) -> List[str]:
+        """Run build process.
+
+        In managed-mode (and eventually destructive-mode), build for each bases
+        configuration which has a matching build-on to the host we are executing
+        on.  Warn for each bases configuration that is unbuildable.  Error if
+        unable to produce any builds for any bases configuration.
+
+        Otherwise, build the charm in the legacy mode until we have complete
+        provider support and can drop it.
+
+        :returns: List of charm files created.
+        """
+        charms: List[str] = []
+
+        if is_charmcraft_running_in_managed_mode():
+            if self.config.bases is None:
+                # XXX: 2021-06-16 Patterson temporary until we have the default base
+                raise CommandError("Bases are currently required in managed-mode.")
+
+            for i, bases_config in enumerate(self.config.bases):
+                for j, build_on in enumerate(bases_config.build_on):
+                    matches, reason = check_if_base_matches_host(build_on)
+                    if matches:
+                        logger.debug(
+                            f"Building for 'bases[{i}]' as host matches 'build-on[{j}]'."
+                        )
+                        charm_name = self.build_charm(bases_config)
+                        charms.append(charm_name)
+                        break
+                    else:
+                        logger.debug(
+                            f"Host does not match 'bases[{i}].build-on[{j}]' ({reason})"
+                        )
+                else:
+                    logger.warning(
+                        f"No suitable 'build-on' environment found in 'bases[{i}]' configuration."
+                    )
+
+            if not charms:
+                raise CommandError(
+                    "No suitable 'build-on' environment found in any 'bases' configuration."
+                )
+
+        else:
+            charm_name = self.build_charm(None)
+            charms.append(charm_name)
+
+        return charms
```

Comment:
Currently this list will always have one entry only, but you're using a list here because in the future (multiple bases in the config, using managed environments) you actually will get multiple entries, right?

Reply:
It can return multiple now, see https://github.com/canonical/charmcraft/pull/390/files#diff-cdf2ae508434a0f8c34dfa438ccb1e70776b6a499ad71e4dea2f20fbf2e15358R555

The next step will be to add `--bases-index` which will be used to constrain it to one result in real-world usage.  But for now, the behavior matches the anticipated destructive-mode.

Reply:
ah, right! Great